### ISSUE TITLE
Haskell backend not fully evaluating predicate

### ISSUE THREAD

**Issue by virgil-serbanuta** [kframework/kore#2684]
The following part of a predicate is not fully evaluated. This started happening at commit https://github.com/kframework/kore/commit/74d28e09382cb0206cb8e087dc172cd5539b7e68 .

```
...
#And
  {
    true
  #Equals
    ActionIndex0 <=Int ActionLastIndex0
  }
...
#And
  {
    u ( ActionLastIndex0 +Int 1 ) in_keys ( _ActionData )
  #Equals
    ( u ( ActionLastIndex0 +Int 1 ) ~> . ) ==K ( u ( ActionIndex0 ) ~> . )
  }
```

To reproduce:

Unpack [bug.tar.gz](https://github.com/kframework/kore/files/6676686/bug.tar.gz)
```
cd bug
kore-repl \
    --repl-script kast.kscript \
    --smt-timeout 4000 \
    proof-board-members-sign-for-kompile.definition.kore \
    --prove proof-board-members-sign-for-kompile.spec.kore \
    --module PROPERTIES-EXECUTE \
    --spec-module PROOF-BOARD-MEMBERS-SIGN-FOR \
    --output proof-board-members-sign-for.output.k
```

In the repl:
```
Kore (0)> konfig
```

**Comment by ana-pantilie** [comment]
What should this evaluate to?

**Comment by ttuegel** [comment]
This is not something that the backend should be able to evaluate. It is plausible that Z3 could show that this is infeasible, but the `in_keys` complicates matters. It's also possible that this state is infeasible for some reason not shown here. @virgil-serbanuta Could you show the entire output?

**Comment by virgil-serbanuta** [comment]
Sure, but first let me say that this issue started occurring at an earlier commit, most likely this one (though it can't be reproduced with the spec above, see below for another example): https://github.com/kframework/kore/commit/67d4d23698551ba29228864fdff55a15d86f3d55

Also, note that this seems to be a rather drastic change in the Haskell backend behaviour.

The new example: [bug1.tar.gz](https://github.com/kframework/kore/files/6678013/bug1.tar.gz)
To reproduce:
```
kore-repl --repl-script kast.kscript --smt-timeout 4000 proof-perform-action-endpoint-add-board-member-Proposer-kompile.definition.kore --prove proof-perform-action-endpoint-add-board-member-Proposer-kompile.spec.kore  --module FUNCTIONS-EXECUTE --spec-module PROOF-PERFORM-ACTION-ENDPOINT-ADD-BOARD-MEMBER-PROPOSER --output proof-perform-action-endpoint-add-board-member-Proposer.output.k

konfig
```

Here are the changes relative to the previous commit, with comments about what should have happened:

* This should be present in the predicate because `_AddressToUserId address ( CallerAddress0 ) |-> u ( CallerId0 ) address ( UserAddress0 ) |-> u ( UserId0 )` should be defined:
```
  #Not ( {
    CallerAddress0
  #Equals
    UserAddress0
  } )
```

* `invariantStateFull` should be evaluated.

* The predicate contains several instances of `( address ( CallerAddress0 ) ~> . ) ==K ( address ( UserAddress0 ) ~> . )` that cannot be evaluated because of the first change in this list.

* Although this exists in the predicate, both before and at the commit above:
```
#And
  #Not ( {
    CallerId0
  #Equals
    UserId0
  } )
```
only before the commit the second term in the following equality is evaluated to `false`:
```
  {
    u ( UserId0 ) in_keys ( UserIdToRoleInner )
  #Equals
    ( u ( UserId0 ) ~> . ) ==K ( u ( CallerId0 ) ~> . )
  }
```

<details>
  <summary>Here is the full state at commit 67d4d23698551ba29228864fdff55a15d86f3d55</summary>

  ```
  #Not ( {
    CallerId0
  #Equals
    0
  } )
#And
  #Not ( {
    CallerId0
  #Equals
    UserId0
  } )
#And
  #Not ( {
    Role
  #Equals
    None
  } )
#And
  #Not ( {
    UserId0
  #Equals
    0
  } )
#And
  <T>
    <TT>
      <k>
        call ( performActionEndpoint ( u ( ActionId0 ) ) ) ~> K ~> .
      </k>
      invariantStateFull ( u ( NumUsers0 ) , UserIdToAddress , _AddressToUserId
      address ( CallerAddress0 ) |-> u ( CallerId0 )
      address ( UserAddress0 ) |-> u ( UserId0 ) , u ( NumBoardMembers ) , u ( NumProposers ) , UserIdToRoleInner
      u ( CallerId0 ) |-> Role:UserRole
      u ( UserId0 ) |-> Proposer , u ( Quorum ) , u ( ActionLastIndex0 ) , ActionData
      u ( ActionId0 ) |-> AddBoardMember ( address ( UserAddress0 ) ) , ActionSigners , address ( CallerAddress0 ) , Stack , .Map , PerformedActions )
    </TT>
  </T>
#And
  {
    ( address ( CallerAddress0 ) ~> . ) ==K ( address ( UserAddress0 ) ~> . ) orBool ( address ( CallerAddress0 ) in_keys ( _AddressToUserId ) )
  #Equals
    ( address ( CallerAddress0 ) ~> . ) ==K ( address ( UserAddress0 ) ~> . )
  }
#And
  {
    address ( CallerAddress0 ) in_keys ( _AddressToUserId )
  #Equals
    ( address ( CallerAddress0 ) ~> . ) ==K ( address ( UserAddress0 ) ~> . ) orBool ( address ( CallerAddress0 ) in_keys ( _AddressToUserId ) )
  }
#And
  {
    false
  #Equals
    address ( UserAddress0 ) in_keys ( _AddressToUserId )
  }
#And
  {
    false
  #Equals
    u ( ActionId0 ) in_keys ( ActionData )
  }
#And
  {
    false
  #Equals
    u ( CallerId0 ) in_keys ( UserIdToRoleInner )
  }
#And
  {
    true
  #Equals
    ( Role:UserRole ~> . ) ==K ( BoardMember ~> . ) orBool ( Role:UserRole ~> . ) ==K ( Proposer ~> . )
  }
#And
  {
    true
  #Equals
    Quorum <=Int boolToInt ( ( Role:UserRole ~> . ) ==K ( BoardMember ~> . ) andBool #listContains ( toExpressionListOrDefault ( ActionSigners [ u ( ActionId0 ) ] orDefault [ . ] , [ void , . ] ) , u ( CallerId0 ) ) ) +Int countCanSignFunction ( toExpressionListOrDefault ( ActionSigners [ u ( ActionId0 ) ] orDefault [ . ] , [ void , . ] ) , UserIdToRoleInner )
  }
#And
  {
    true
  #Equals
    actionSignersInvariant ( ActionSigners )
  }
#And
  {
    true
  #Equals
    addressToUserIdInvariant ( _AddressToUserId )
  }
#And
  {
    true
  #Equals
    userIdToRoleInvariant ( UserIdToRoleInner )
  }
#And
  {
    true
  #Equals
    valueNotInMapValues ( u ( CallerId0 ) , _AddressToUserId )
  }
#And
  {
    true
  #Equals
    valueNotInMapValues ( u ( UserId0 ) , _AddressToUserId )
  }
#And
  {
    u ( UserId0 ) in_keys ( UserIdToRoleInner )
  #Equals
    ( u ( UserId0 ) ~> . ) ==K ( u ( CallerId0 ) ~> . )
  }
  ```  
</details>

<details>
  <summary>Here is the full state before that commit</summary>

  ```
  #Not ( {
    CallerAddress0
  #Equals
    UserAddress0
  } )
#And
  #Not ( {
    CallerId0
  #Equals
    0
  } )
#And
  #Not ( {
    CallerId0
  #Equals
    UserId0
  } )
#And
  #Not ( {
    Role
  #Equals
    None
  } )
#And
  #Not ( {
    UserId0
  #Equals
    0
  } )
#And
  <T>
    <TT>
      <k>
        call ( performActionEndpoint ( u ( ActionId0 ) ) ) ~> K ~> .
      </k>
      <state>
        <multisig-state>
          <users>
            <num-users>
              u ( NumUsers0 )
            </num-users>
            <user-id-to-address>
              UserIdToAddress
            </user-id-to-address>
            <address-to-user-id>
              _AddressToUserId
              address ( CallerAddress0 ) |-> u ( CallerId0 )
              address ( UserAddress0 ) |-> u ( UserId0 )
            </address-to-user-id>
          </users>
          <board-state>
            <num-board-members>
              u ( NumBoardMembers )
            </num-board-members>
            <num-proposers>
              u ( NumProposers )
            </num-proposers>
            <user-roles>
              UserIdToRoleInner
              u ( CallerId0 ) |-> Role:UserRole
              u ( UserId0 ) |-> Proposer
            </user-roles>
            <quorum>
              u ( Quorum )
            </quorum>
          </board-state>
          <action-state>
            <action-last-index>
              u ( ActionLastIndex0 )
            </action-last-index>
            <actions>
              <action-data>
                ActionData
                u ( ActionId0 ) |-> AddBoardMember ( address ( UserAddress0 ) )
              </action-data>
              <action-signers>
                ActionSigners
              </action-signers>
            </actions>
          </action-state>
        </multisig-state>
        <pseudocode-state>
          <variables>
            .Map
          </variables>
          <stack>
            Stack
          </stack>
        </pseudocode-state>
        <external-call-env>
          <caller-address>
            address ( CallerAddress0 )
          </caller-address>
        </external-call-env>
        <log>
          <performed-actions>
            PerformedActions
          </performed-actions>
        </log>
      </state>
    </TT>
  </T>
#And
  {
    false
  #Equals
    address ( CallerAddress0 ) in_keys ( _AddressToUserId )
  }
#And
  {
    false
  #Equals
    address ( UserAddress0 ) in_keys ( _AddressToUserId )
  }
#And
  {
    false
  #Equals
    u ( ActionId0 ) in_keys ( ActionData )
  }
#And
  {
    false
  #Equals
    u ( CallerId0 ) in_keys ( UserIdToRoleInner )
  }
#And
  {
    false
  #Equals
    u ( UserId0 ) in_keys ( UserIdToRoleInner )
  }
#And
  {
    true
  #Equals
    ( Role:UserRole ~> . ) ==K ( BoardMember ~> . ) orBool ( Role:UserRole ~> . ) ==K ( Proposer ~> . )
  }
#And
  {
    true
  #Equals
    Quorum <=Int boolToInt ( ( Role:UserRole ~> . ) ==K ( BoardMember ~> . ) andBool #listContains ( toExpressionListOrDefault ( ActionSigners [ u ( ActionId0 ) ] orDefault [ . ] , [ void , . ] ) , u ( CallerId0 ) ) ) +Int countCanSignFunction ( toExpressionListOrDefault ( ActionSigners [ u ( ActionId0 ) ] orDefault [ . ] , [ void , . ] ) , UserIdToRoleInner )
  }
#And
  {
    true
  #Equals
    actionSignersInvariant ( ActionSigners )
  }
#And
  {
    true
  #Equals
    addressToUserIdInvariant ( _AddressToUserId )
  }
#And
  {
    true
  #Equals
    userIdToRoleInvariant ( UserIdToRoleInner )
  }
#And
  {
    true
  #Equals
    valueNotInMapValues ( u ( CallerId0 ) , _AddressToUserId )
  }
#And
  {
    true
  #Equals
    valueNotInMapValues ( u ( UserId0 ) , _AddressToUserId )
  }
  ```  
</details>

**Comment by ttuegel** [comment]
@ana-pantilie Here is a key clause I think we should consider:

```
{
  address ( CallerAddress0 ) in_keys ( _AddressToUserId )
#Equals
  ( address ( CallerAddress0 ) ~> . ) ==K ( address ( UserAddress0 ) ~> . ) orBool ( address ( CallerAddress0 ) in_keys ( _AddressToUserId ) )
}
#And
{
  false
#Equals
  address ( UserAddress0 ) in_keys ( _AddressToUserId )
}
```

Why isn't the second equality used to generate a substitution into the first? That is necessary for simplification to proceed.

**Comment by ana-pantilie** [comment]
> Why isn't the second equality used to generate a substitution into the first?

How? The second equality contains `address ( UserAddress0 ) in_keys ( _AddressToUserId )`, not `address ( CallerAddress0 ) in_keys ( _AddressToUserId )` so I don't understand how that would work.

**Comment by ttuegel** [comment]
You're right, I think I mis-read that term.

So, I actually can't think of any reason that the backend should be able to simplify this further, and I think that the "proof" using the earlier commit is incorrect.

**Comment by virgil-serbanuta** [comment]
@ttuegel While the previous proof may be incorrect, please note that `invariantStateFull` should probably be evaluated regardless of what happens to the predicate. Also, consider this:

The backend can, in principle, automatically add `#Ceil(LHS)` to the predicate. I think it used to do that, maybe it still does. When simplified, `#Ceil(LHS)` should contain, among others `#Ceil(_AddressToUserId   address ( CallerAddress0 ) |-> u ( CallerId0 )   address ( UserAddress0 ) |-> u ( UserId0 ))`. Obviously, this implies the following two terms, which, somehow, used to be part of the predicate:

```
  address ( CallerAddress0 ) in_keys ( _AddressToUserId ) #Equals false
#And
  #Not( CallerAddress0 #Equals UserAddress0 )
```

I don't know why this is not happening anymore, but a proof can't be incorrect just because it relies on terms which are implied by `#Ceil(LHS)`.

**Comment by ana-pantilie** [comment]
Why can't `invariantStateFull` be evaluated now?

**Comment by virgil-serbanuta** [comment]
FWIW: If I define a new function `not_in_keys(KItem, Map)` and I use that instead of  `in_keys` everywhere, this issue seems to go away (`invariantStateFull` is evaluated, and the predicate looks more or less as expected).

I think that this happens because the `orBool` expression mentioned here: https://github.com/kframework/kore/issues/2684#issuecomment-864426036  (which is produced by a custom `in_keys` simplification rule) becomes `andBool` for the equivalent `not_in_keys` simplification rule, and the Haskell backend finds it easier to deal with `andBool` than with `orBool`. If that's, indeed, the case, then it would be nice if the Haskell backend would simplify, say, `false #Equals (X orBool Y)` to `true #Equals ((notBool X) andBool (notBool Y))`, which would then become `(true #Equals notBool X) #And (true #Equals notBool Y)`.

Returning to `not_in_keys`, I'm worried that I'll have performance problems if I'm using a user-defined function instead of a builtin.

**Comment by ttuegel** [comment]
> it would be nice if the Haskell backend would simplify, say, `false #Equals (X orBool Y)`

The backend simplifies `false #Equals (X orBool Y)` to `{ false #Equals X } #And { false #Equals Y }`.

**Comment by ana-pantilie** [comment]
I wonder if this was fixed in f67a42d9b95ae92fd683da3dab464dd107817b56 (https://github.com/kframework/kore/pull/2714)?

**Comment by virgil-serbanuta** [comment]
Yes, it seems so. Thanks!
